### PR TITLE
Orchestra non-storing rule update

### PR DESCRIPTION
**Pull request by atiselsts**
This PR improve the energy consumption of the non-storing (NS) rule by only adding Tx links in the timeslots that have a corresponding active neighbor. The idea is outlined and discussed here https://github.com/contiki-ng/contiki-ng/issues/889#issuecomment-566695431

There is a corresponding change in the IPv6 DS code: a new pair of callbacks are added: NETSTACK_CONF_DS6_NEIGHBOR_ADDED_CALLBACK and NETSTACK_CONF_DS6_NEIGHBOR_REMOVED_CALLBACK. These callbacks are only defined when Contiki-NG is built with the Orchestra scheduler.

### CODE REVIEW

**Comment by atiselsts** [comment]
For now this is in draft state, as  #1518 should be merged first, and then this one updated.

**Comment by atiselsts** [comment]
Now that the root rule PR is merged, this one is ready for review!

**Comment by atiselsts** [comment]
Probably needs a more extensive review, makes sense to postpone this to future versions IMO.

**Review by arurke — CHANGES_REQUESTED** [review]
Looks good @atiselsts, just one substantial comment/question, rest is mostly code style. I was thinking to do a quick sanity in FIT IoT-lab, but imo these changes are pretty straightforward.

File: os/services/orchestra/orchestra-rule-unicast-per-neighbor-rpl-ns.c [hunk 1]
```diff
@@ -69,13 +71,42 @@ get_node_channel_offset(const linkaddr_t *addr)
...
+
+  if(linkaddr != NULL) {
+
+    timeslot = get_node_timeslot(linkaddr);
+
+    /* Add a Tx link to the neighbor; do not replace any existing links at that cell */
+    tsch_schedule_add_link(sf_unicast,
+        LINK_OPTION_SHARED | LINK_OPTION_TX,
+        LINK_TYPE_NORMAL, &tsch_broadcast_address,
+        timeslot, 0, 0);
```
[the remark this answers is not in the document]
Comment (atiselsts) [reply os/services/orchestra/orchestra-rule-unicast-per-neighbor-rpl-ns.c:86]: The channel offset here should not matter: select_packet() always sets the right channel offset per packet. I will add a comment that explains this.

[the remark this answers is not in the document]
Comment (arurke) [reply os/services/orchestra/orchestra-rule-unicast-per-neighbor-rpl-ns.c:85]: Ah, of course, thx for the clarification

**Comment by arurke** [comment]
:+1: Thanks for the changes and explanation. You want to squash the last commit before I merge?

**Comment by atiselsts** [comment]
Commit squashed!

**Review by arurke — APPROVED** [review]

**Pull request merged**